### PR TITLE
cAVS IPC driver, power & clock unit tests

### PR DESCRIPTION
**Pull request by andyross**
This adds a register interface and driver for the host IPC hardware (essentially the bidirectional interrupt trigger with the host PC), along with a test for it based on some rework of the python scripting submitted earlier in #41803 (which I'm about to close).

### CODE REVIEW

**Review by teburd — APPROVED** [review]

File: soc/xtensa/intel_adsp/common/cavs_ipc.c
Comment (andyross) [reply soc/xtensa/intel_adsp/common/cavs_ipc.c:150]: No, that's how I had it.  It didn't build.  Frankly it doesn't matter for these platforms since dynamic IRQs are architectural, but something to investigate.

**Comment by andyross** [comment]
Sneak in one more patch with a productized smoke test for each core's cache & clock state.

File: soc/xtensa/intel_adsp/common/cavs_ipc.c
Comment (teburd) [reply soc/xtensa/intel_adsp/common/cavs_ipc.c:150]: Is this with the xtensa compiler in the sdk or with the proprietary job? worked fine for me with the zephyr sdk compiler as I wrote it. I just built the board test with it.

Comment (andyross) [reply soc/xtensa/intel_adsp/common/cavs_ipc.c:150]: It was just the SDK.  I'll go back and check, it's entirely possible I messed something up.  I have minimal patience for devicetree burps, to be perfectly honest.

File: include/drivers/cavs_ipc.h
Comment (andyross) [inline include/drivers/cavs_ipc.h:26]: Hm.... I think I dropped a change here somewhere.  Somewhere I wrote a note about the fact that async DONE interrupts on cAVS 1.5 don't actually work (sorta, it's complicated), but I don't see it in the code as pushed.  Will fix.

File: soc/xtensa/intel_adsp/common/cavs_ipc.c
Comment (nashif) [inline soc/xtensa/intel_adsp/common/cavs_ipc.c:1]: why put the driver under the SOC and not under drivers/ipc/....

File: soc/xtensa/intel_adsp/common/include/cavs-ipc-regs.h
Comment (nashif) [inline soc/xtensa/intel_adsp/common/include/cavs-ipc-regs.h:12]: what is ACE?

**Review by nashif — DISMISSED** [review]

File: soc/xtensa/intel_adsp/common/cavs_ipc.c
Comment (andyross) [reply soc/xtensa/intel_adsp/common/cavs_ipc.c]: Turned out that it didn't like the left typecast in the ISR parameter, because it wants to use it to glue up a symbol name.   Fixed.

File: include/drivers/cavs_ipc.h
Comment (andyross) [reply include/drivers/cavs_ipc.h]: It can indeed, fixed

File: soc/xtensa/intel_adsp/common/cavs_ipc.c
Comment (andyross) [reply soc/xtensa/intel_adsp/common/cavs_ipc.c:1]: It's a SOC-specific device for now.  I'll see about adding an extra API layer above it that can be more portably implemented

**Comment by andyross** [comment]
Seems like this may be in review for another few days, so sneak in three more patches containing work on the MP power management and some smoke test additions

**Review by dcpleung — DISMISSED** [review]

File: soc/xtensa/intel_adsp/common/cavs_ipc.c
Comment (lgirdwood) [reply soc/xtensa/intel_adsp/common/cavs_ipc.c:1]: ack - this needs to SOC and protocol generic so it can be used by any SOC. At at mid level we need something like (feel free to modify)
```c
int ipc_send_message(uint32_t addr, char *doorbell, size_t door_size, char *mailbox, size_t mail_size, int blocking);
int ipc_recv_message(uint32_t addr, int (callback_func*) (doorbell, door_size, mailbox, mailsize));
```
The higher level application API would in cSOF would be similar to what it is today, but it would do the protocol mapping and call the above zephyr APIs.

Comment (andyross) [reply soc/xtensa/intel_adsp/common/cavs_ipc.c:1]: Inbound in maybe an hour.  Really it just productizes the "simple" bits of this hardware (so no weird return/DONE interrupt) as a generic subsystem API and implements it with the device as it stands already.  Finishing the typing part, need to write a test and it'll go onto the end of this PR.

**Comment by andyross** [comment]
Push a more general API in the final patch of the series.  All this does is add a Zephyr device model subsystem wrapper around the driver that exposes the minimal/simplest/most-easily-reimplementable-on-other-platforms subset[1] of the IPC functionality as a generic API called "inotice".[2]

So instead of using the underlying cavs_ipc_*() API, you can use inotice_*() wrappers which indirect through the .api pointer of the device to find the specific callbacks.  And I was pleased to see that as long as one is careful to use "INOTICE_DEV" (which is a macro that expands to the static devicetree-instantiated device object), the wrappers get folded by the optimizer and the generated code is identical.

[1] Basically just sending messages consisting of two dwords of data (with platform-dependent significant bits) and receiving them on the other side in an interrupt callback.  I dropped the DONE interrupts and async command completion, though because the underlying cavs_ipc driver is still there you can use those features directly on the same device pointer and they'll work fine, just not portably.

[2] I'm not calling it IPC.  "Inter-process communication" in my world mean semaphores and queues and pipes and locks.  They aren't a hardware device.

File: soc/xtensa/intel_adsp/tools/cavstool.py
Comment (nashif) [inline soc/xtensa/intel_adsp/tools/cavstool.py:319]: Need some error checking here and some help text describing what this tool does, if I run it standalone without arguments, I get:
```
Traceback (most recent call last):
  File "/home/nashif/cavstool.py", line 336, in <module>
    asyncio.get_event_loop().run_until_complete(main())
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/nashif/cavstool.py", line 319, in main
    load_firmware(sys.argv[1])
IndexError: list index out of range
```

and it needs root, so without sudo I get:

```
ups1:zephyr(main): python3 ~/cavstool.py
Traceback (most recent call last):
  File "/home/nashif/cavstool.py", line 336, in <module>
    asyncio.get_event_loop().run_until_complete(main())
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/home/nashif/cavstool.py", line 316, in main
    (hda, sd, dsp, hda_ostream_id, cavs15) = map_regs()
  File "/home/nashif/cavstool.py", line 38, in map_regs
    with open(f"{pcidir}/power/control", "w") as ctrl:
PermissionError: [Errno 13] Permission denied: '/sys/bus/pci/devices/0000:00:0e.0/power/control'

```

File: include/drivers/cavs_ipc.h
Comment (lgirdwood) [inline include/drivers/cavs_ipc.h:9]: Is there any need for a drivers/cavs_ipc.h as I would expect the inotice header to define the operations that each SOC IPC driver would use ?

Comment (andyross) [reply include/drivers/cavs_ipc.h:9]: cavs_ipc is a more featureful interface, with support for async "DONE" notification on command completion.  The driver supports the inotice interface, which is a subsystem that can be implemented portably by any driver.  But also has its own API for the more complicated stuff.

Is that "needed"?  I don't know, but the hardware is there and it works and we have a driver and a test.  Seems like we should probably leave it in if someone wants to be creative with this device in the future.  You could use it for talking to the CSME if someone wants to evolve the protocol, etc...

Alternatively we could move the header back into the soc include directory so it doesn't confuse anyone.

File: soc/xtensa/intel_adsp/tools/cavstool.py
Comment (andyross) [reply soc/xtensa/intel_adsp/tools/cavstool.py:319]: This got pulled out and submitted separately in #42020 where that polish got added.

File: include/drivers/cavs_ipc.h
Comment (andyross) [reply include/drivers/cavs_ipc.h:9]: Oh, and to be clear there is no overhead: the functions that are common between the interfaces are identical, the calls from cavs_ipc that are unused get optimized out.  The ISR is identical regardless.

**Comment by andyross** [comment]
Rebase; the cavstool script merged separately.

**Comment by andyross** [comment]
So how is this looking?  Things are starting to back up behind it, is this something folks think can be merged?  If not, can I get some direction as to what needs to be changed?  If we can't get to a resolution on an appropriate abstraction, can we at least merge the hardware driver layer on top of which we (really SOF: it's SOF that needs the abstraction) can build something later?

**Comment by andyross** [comment]
It seems like we're in a sort of paralysis loop here, with @lgirdwood asking for an abstracted AP, and @nashif and @MaureenHelm uncomfortable with my (admittedly minimal) existing attempt.

But... there are some practical concerns here:

1. I need to get this merged so I can continue submitting more work

2. I need **this driver** merged, or something very like it, because I'm now using the host communication to the loader script to be able to write unit tests for multiprocessor startup and clock control.  And we want those!

So, how about this: I remove the cavs_ipc.h header and inotice.h headers both from the public include directories and put them in as SOC-level APIs only.  I can use them from there, and in the future when we decide on an appropriately portable SOF interface we can either evolve it or implement the portable API in terms of cavs_ipc.

Will that unblock the merge here?

**Comment by andyross** [comment]
As promised, here's a version without any new public APIs, limiting the cavs_ipc driver to a soc role.  And I added the patches that are now using it to effect unit testing on the hardware, to demonstrate the value of getting this merged independently of a design for a portable SOF abstraction.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Excellent looking series, really good new coverage to the tests!

Unfortunately a test run with SOF caught an error. I did a quick bisect and basicly "kernel: Refactor SMP cpu initialization a bit" breaks SOF test cases on cAVS2.5 using the second DSP core. Launch CPU fails. I'll try to debug this a bit more today.

UPDATE: a potential SOF side change is needed after this PR. I'll update later on this. If yes, we can proceed with the merge and do matching change on SOF side.

**Review by kv2019i — DISMISSED** [review]
Ok to go ahead with this. SOF has to make a matching patch when this is merged (for 3.1), but this is ok.

**Review by teburd — DISMISSED** [review]
Re-reviewing as this has change quite a lot in the 19 or so days since I last looked. This is a nice changeset, and I'll be needing something (this exact sort of IPC with cavstool.py changes to test the HDA streams soon.

**Comment by teburd** [comment]
I now have HDA working with host in/out streams and all, it depends on this branch currently as I'm using the cavs_ipc.h mechanism to ensure proper ordering of operations between the host and dsp. Any updates on this?

**Comment by andyross** [comment]
> Why wont we start using more explicit/descriptive include paths

No argument at all from me about where to put the header.  This one is where it is simply because it's a SOC-specific driver (i.e. in its own directory that doesn't exist on Zephyr builds for other platforms) and because "cavs" is in the filename already so it seemed like it was reasonably disambiguated.  For cross platform driver subsystems (like IPM), we do in fact have a curated tree in include/drivers/<subsys>/...

Mostly though, these discussions generally turn into big bikesheds, so I keep my head down and try to structure changes to have minimal impact.  Realistically we're always moving stuff around anyway so any decision made will be ephemeral anyway. :)

But sure: if someone wants the driver somewhere else just say so and I'll move it.

File: soc/xtensa/intel_adsp/common/cavs_ipc.c
Comment (andyross) [reply soc/xtensa/intel_adsp/common/cavs_ipc.c:30]: Yes, if we want to use the cavs_ipc driver on top of non-host IPC blocks that share interrupts with each other.  There, you need to figure out what device is involved via some other means and then call into the cavs_ipc "ISR" to do the per-IPC work.

File: soc/xtensa/intel_adsp/common/cavs_ipc.c
Comment (andyross) [reply soc/xtensa/intel_adsp/common/cavs_ipc.c:122]: Synchronization right here is actually provided by the spinlock.  The semaphore is for blocking  (exactly) one thread in send_message_sync().

File: soc/xtensa/intel_adsp/common/mp_cavs.c
Comment (andyross) [reply soc/xtensa/intel_adsp/common/mp_cavs.c:26]: It's provided by the OS now, arch_start_cpu() is unsynchronized.

File: tests/boards/intel_adsp/smoke/src/cpus.c
Comment (andyross) [reply tests/boards/intel_adsp/smoke/src/cpus.c:17]: Actually yes.  Or I've seen that happen historically (no idea if Zephyr/SOFs current toolchains and compile arguments produce that behavior).  And in any case, they're absolutely allowed to.  Volatile guarantees correct behavior.

**Comment by andyross** [comment]
@lyakh I'm sorry, I came by to push another version and realized there were a bunch of stale comments to your review.  Please let me know if these don't resolve your concerns.  New version inbound in 3, 2, 1...

**Review by andyross — COMMENTED** [review]
oops still didn't flush comments

**Comment by andyross** [comment]
OK, new version up.  @lgirdwood what you will want to look at are the final three patches, which puts the SOF ipc3/ipc4 protocols under a Zephyr "IPM" device.  See API at https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/ipm.h (or generated doxygen in a link I'm having trouble finding at the moment, of course).

@teburd Your DMA PR on top of this is likely to collide a tiny bit (if nothing else with the command IDs for the cavstool script, because I added a new one), sorry.   Let me know if you have any trouble.

@aborisovich The broad idea here (that I **think** will work, but again this is based on merely reading the SOF ipc layer) is that you replace the platform_* layer under ipc3/4 with wrappers on this IPM device.  The first word of the command (the one passed in the TDR/IDR registers) goes in the IPM "id" parameter, the remaining ones are just passed as a byte array.  The new asynchronous mode added means that you can hold onto the received buffer (which is always the same inbox memory on cavs) and leave it in place as the result of mailbox_validate().  I do know ipc4 wants to leave this in place during processing (where ipc3 would copy it).  Hopefully there's no major surgery needed to SOF, just some wrappers.  And ideally there should be no changes needed to host drivers if I've done it correctly (which is an open question!).  Let me know if anything is confusing.

**Comment by teburd** [comment]
No worries, I knew you were improving it, my patchset is relatively small and should be easy enough to rebase, will look closer at this soon

**Comment by andyross** [comment]
Sorry, the push I just did was to clean up some missing licenses and add a CODEOWNERS line for the new driver.  I promise I'll yell if there are logic changes. :)

File: include/drivers/ipm.h
Comment (teburd) [inline include/drivers/ipm.h:249]: Can I recommend allowing a pass through void* userdata as well here if that seems sensible, otherwise you are forcing static data to be used and you may wish to pass in a ipc object alongside the callback here without always creating custom callbacks.

**Review by teburd — DISMISSED** [review]
Looks good to me.

File: include/drivers/ipm.h
Comment (andyross) [reply include/drivers/ipm.h:249]: To be clear: that function pointer is a method, it's implemented by the driver, not the user of the API.

Comment (teburd) [reply include/drivers/ipm.h:249]: Bah, ignore then, I commented too quickly on this

**Review by dcpleung — APPROVED** [review]

**Review by teburd — APPROVED** [review]

File: tests/boards/intel_adsp/smoke/src/cpus.c
Comment (marc-hb) [reply tests/boards/intel_adsp/smoke/src/cpus.c:17]: I remember having to find the optimizer every single time I tried to add code that "does nothing" (fortunately not that often).

On the other hand I'm surprised the test framework does not already have something like this, is this sort of test new to Zephyr?

**Review by nashif — APPROVED** [review]

**Pull request merged**
